### PR TITLE
OCPCLOUD-2565: Adds Paused condition for Machines and MachineSets

### PR DESCRIPTION
**Pull request by theobarberbany**
This adds a Paused Condition for both Machines and MachineSets. 

It sets the paused condition false when the `Machine` or `MachineSet` status has
the `AuthoritativeAPI` set to `MachineAPI`, otherwise it sets it to true and
stops reconciliation.

### CODE REVIEW

File: pkg/controller/machine/controller.go [hunk 1]
```diff
@@ -167,6 +173,41 @@ func (r *ReconcileMachine) Reconcile(ctx context.Context, request reconcile.Requ
...
+	// Check Status.AuthoritativeAPI
+	// If not MachineAPI. Set the paused condition true and return early.
+
+	// @joelspeed i think we want some defaulting here, i.e if unset then default
+	// to MAPI, as it's not currently set. not sure if that's part of another
+	// card for setting the status / looking at the migration logic or best done
+	// here. Currently, if the field is unset we will set the paused condition
+	// and not reconcile.
+
+	if m.Status.AuthoritativeAPI != machinev1.MachineAuthorityMachineAPI {
```
Comment (theobarberbany) [inline pkg/controller/machine/controller.go:185]: I'm not sure if this is the exact behaviour we want. Currently if the `AuthoritativeAPI` is unset we will pause and not reconcile. 

I figure 'i'm not sure so do nothing' is an OK approach, but let me know if the plan is different :)

Comment (JoelSpeed) [reply pkg/controller/machine/controller.go:178]: The card says:
```
* Should not reconcile when .status.authoritativeAPI is not MachineAPI
* Except when it is empty (prior to defaulting migration webhook)
```
Also, note, the feature work here should be gated on the presence of the feature gate being enabled, we need to work that in here as well

File: pkg/controller/machine/controller.go [hunk 2]
```diff
@@ -71,6 +71,12 @@ const (
 	unknownInstanceState = "Unknown"
 
 	skipWaitForDeleteTimeoutSeconds = 1
+
+	PausedCondition machinev1.ConditionType = "Paused"
+
+	PausedConditionReason = "AuthoritativeAPI is not set to MachineAPI"
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller.go:77]: Reasons are typically short PamelCase strings, so I'd expect, `AuthoritativeAPINotMachineAPI`

File: pkg/controller/machine/controller.go [hunk 3]
```diff
@@ -167,6 +173,34 @@ func (r *ReconcileMachine) Reconcile(ctx context.Context, request reconcile.Requ
...
+	// Set the paused condition to false, continue reconciliation
+	conditions.Set(m, conditions.FalseCondition(
+		PausedCondition,
+		NotPausedConditionReason,
+		machinev1.ConditionSeverityInfo,
+		"The AuthoritativeAPI is set to %s", m.Status.AuthoritativeAPI,
+	))
+	if patchErr := r.updateStatus(ctx, m, ptr.Deref(m.Status.Phase, ""), nil, originalConditions); patchErr != nil {
+		klog.Errorf("%v: error patching status: %v", machineName, patchErr)
+	}
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller.go:201]: Is there an update status later in the reconcile? Or one that is deferred? Wondering if this could increases our 409 count if we add additional updates mid flow.

Would ideally expect at most one status update per reconcile

File: pkg/controller/machineset/machineset_controller_test.go [hunk 4]
```diff
@@ -131,6 +140,71 @@ var _ = Describe("MachineSet Reconciler", func() {
...
+
+		By("Verifying the paused condition is approproately set to true")
+		Eventually(func() bool {
+			if err := k8sClient.Get(ctx, objectKey(instance), instance); err != nil {
+				return false
+			}
+
+			// The condition is set to true
+			return conditions.IsTrue(instance, PausedCondition)
+		}, timeout*10).Should(BeTrue())
```
Comment (JoelSpeed) [inline pkg/controller/machineset/machineset_controller_test.go:187]: If you just returned the conditions objects, this may print better errors when there's an issue
```suggestion
		Eventually(func() ([]machinev1.Condition, error) {
			if err := k8sClient.Get(ctx, objectKey(instance), instance); err != nil {
				return nil, err
			}

			// The condition is set to true
			return instance.Status.Conditions, nil
		}, timeout*10).Should(ContainElement(SatisfyAll(
			HaveField("Type", Equal(PausedCondition)),
			HaveField("Status", BeTrue()),
		)))
```

Also, why `timeout*10`?

File: pkg/controller/machine/controller_test.go [hunk 5]
```diff
@@ -1199,3 +1226,123 @@ func TestSetLifecycleHookConditions(t *testing.T) {
 		})
 	}
 }
+
+func TestSetPausedCondition(t *testing.T) {
+	t.Run("PausedCondition sets correctly", func(t *testing.T) {
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller_test.go:1231]: No ginkgo? 😢

File: pkg/controller/machine/controller_test.go [hunk 6]
```diff
@@ -1199,3 +1226,123 @@ func TestSetLifecycleHookConditions(t *testing.T) {
...
+		g.Expect(k8sClient.Create(ctx, namespace)).To(Succeed())
+		defer func() {
+			if err := k8sClient.Delete(ctx, namespace); err != nil {
+				t.Fatalf("error deleting namespace: %v", err)
+			}
+		}()
+
+		// Set up the test machine in a state where we expect the paused condition
+		// to be true
+		machine := &machinev1.Machine{
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller_test.go:1257]: Could use the builders from the cluster-api-actuator-pkg test utils

File: pkg/controller/machine/controller_test.go [hunk 7]
```diff
@@ -1199,3 +1226,123 @@ func TestSetLifecycleHookConditions(t *testing.T) {
...
+		}()
+
+		machine.Status.AuthoritativeAPI = machinev1.MachineAuthorityClusterAPI
+		g.Expect(k8sClient.Status().Update(ctx, machine)).To(Succeed())
+
+		key := objectKey(machine)
+		g.Expect(k8sClient.Get(ctx, key, machine)).To(Succeed())
+		g.Expect(machine.Status.AuthoritativeAPI).To(Equal(machinev1.MachineAuthorityClusterAPI))
+
+		g.Eventually(func() bool {
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller_test.go:1303]: I left another comment on other tests, but, try not to return bools, return structures and use the gomega matchers to extract the data, it's much easier to debug. Any `Eventually().Should(BeTrue())` is an anti-pattern

File: pkg/controller/machine/controller_test.go [hunk 8]
```diff
@@ -1199,3 +1226,123 @@ func TestSetLifecycleHookConditions(t *testing.T) {
...
+			// The condition is set to true
+			return conditions.IsTrue(machine, PausedCondition)
+		}, timeout).Should(BeTrue())
+
+		// Set the AuthoritativeAPI to Migrating as we must transition through it
+		machine.Status.AuthoritativeAPI = machinev1.MachineAuthorityMigrating
+		g.Expect(k8sClient.Status().Update(ctx, machine)).To(Succeed())
+
+		// Do we want to expect we remain paused during migration? Wouldn't this
+		// just instant pass?
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller_test.go:1321]: Yes it would, so you could use a `Consistently` to check that it doesn't get changed

File: pkg/controller/machine/controller_test.go [hunk 9]
```diff
@@ -1199,3 +1226,123 @@ func TestSetLifecycleHookConditions(t *testing.T) {
...
+
+		g.Expect(k8sClient.Create(ctx, machine)).To(Succeed())
+		defer func() {
+			if err := k8sClient.Delete(ctx, machine); err != nil {
+				t.Fatalf("error deleting machine: %v", err)
+			}
+		}()
+
+		machine.Status.AuthoritativeAPI = machinev1.MachineAuthorityClusterAPI
+		g.Expect(k8sClient.Status().Update(ctx, machine)).To(Succeed())
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller_test.go:1297]: Susceptible to 409s?

File: pkg/controller/machine/machine_controller_suite_test.go [hunk 10]
```diff
@@ -37,7 +37,21 @@ var (
...
+		},
+
+		// The CRDs in the CRDDirectoryPaths override the one specified in Paths
+		// above. This breaks tests that rely on parts of the CRD only found in
+		// CustomNoUpgrade. Once the migration APIs land in the default CRD then
+		// this can be uncommented and the Paths above removed.
+
+		//	CRDDirectoryPaths: []string{
+		//		filepath.Join("..", "..", "..", "install"),
+		//	},
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_suite_test.go:54]: ?

File: pkg/controller/machineset/controller.go
Comment (JoelSpeed) [inline pkg/controller/machineset/controller.go:1]: Since this is basically the same as the Machine case, see the comments I left there :)

File: pkg/controller/machineset/machineset_controller_suite_test.go [hunk 11]
```diff
@@ -63,7 +63,23 @@ func TestMachinesetController(t *testing.T) {
...
+
+		// The CRDs in the CRDDirectoryPaths override the one specified in Paths
+		// above. This breaks tests that rely on parts of the CRD only found in
+		// CustomNoUpgrade. Once the migration APIs land in the default CRD then
+		// this can be uncommented and the Paths above removed.
+
+		//	CRDDirectoryPaths: []string{
+		//		filepath.Join("..", "..", "..", "install"),
+		//	},
+
```
Comment (JoelSpeed) [inline pkg/controller/machineset/machineset_controller_suite_test.go:82]: ?

File: pkg/controller/machine/controller.go (hunk 1 above)
Comment (theobarberbany) [reply pkg/controller/machine/controller.go:178]: Ah 🤦  - ok I'll update the code/tests !

File: pkg/controller/machineset/machineset_controller_test.go (hunk 4 above)
Comment (theobarberbany) [reply pkg/controller/machineset/machineset_controller_test.go:187]: > Also, why timeout*10?

Ah - I've left my debugging timeouts in. It was so when I was stepping through there would be no failures due to the timeout being exceeded.

File: pkg/controller/machineset/machineset_controller_suite_test.go (hunk 11 above)
Comment (theobarberbany) [reply pkg/controller/machineset/machineset_controller_suite_test.go:82]: I found that if I left this in, the CRDs synced by the `/hack/crd-sync.sh` script would take priority meaning we wouldnt have the AuthoratitiveAPI field in the status. 

I could not find a way to exclude some files from `CRDDirectoryPaths`

File: pkg/controller/machine/machine_controller_suite_test.go (hunk 10 above)
Comment (theobarberbany) [reply pkg/controller/machine/machine_controller_suite_test.go:54]: See other comment

File: pkg/controller/machine/controller.go (hunk 3 above)
Comment (theobarberbany) [reply pkg/controller/machine/controller.go:201]: - See if we can use existing updates to status rather than introducing a new one.

File: pkg/controller/machine/controller.go [hunk 12]
```diff
@@ -71,6 +71,12 @@ const (
 	unknownInstanceState = "Unknown"
 
 	skipWaitForDeleteTimeoutSeconds = 1
+
+	PausedCondition machinev1.ConditionType = "Paused"
+
+	PausedConditionReason = "AuthoritativeAPINotMachineAPI"
+
+	NotPausedConditionReason = "AuthoritativeAPI is set to MachineAPI"
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller.go:79]: Are these exported for a particular reason?

Comment (JoelSpeed) [inline pkg/controller/machine/controller.go:79]: Reason should be PascalCase, this one is still space separated

File: pkg/controller/machine/controller.go (hunk 1 above)
Comment (JoelSpeed) [reply pkg/controller/machine/controller.go:178]: Don't think this one has been updated?

File: pkg/controller/machine/machine_controller_suite_test.go [hunk 13]
```diff
@@ -18,53 +18,94 @@ package machine
...
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+func init() {
+	textLoggerConfig := textlogger.NewConfig()
+	textLoggerConfig.AddFlags(flag.CommandLine)
+	logf.SetLogger(textlogger.NewLogger(textLoggerConfig))
+
+	// Register required object kinds with global scheme.
+	_ = machinev1.Install(scheme.Scheme)
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_suite_test.go:45]: Why not do this in the BeforeSuite and check the error?

File: pkg/controller/machine/machine_controller_suite_test.go [hunk 14]
```diff
@@ -18,53 +18,94 @@ package machine
...
+	Expect(cfg).ToNot(BeNil())
+
+})
+
+var _ = AfterSuite(func() {
+	By("tearing down the test environment")
+	Expect(testEnv.Stop()).To(Succeed())
+})
+
+func StartEnvTest(t *testing.T) func(t *testing.T) {
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_suite_test.go:88]: Since this is kind of duplicating some of the logic above, is there a way to share them?

Perhaps have the func return the testEnv? And then you can `defer func() { testenv.Stop() }` in the non ginkgo formatted tests?

File: pkg/controller/machine/machine_controller_test.go [hunk 15]
```diff
@@ -17,87 +17,217 @@ limitations under the License.
...
+		k8sClient = mgr.GetClient()
+		k = komega.New(k8sClient)
 
-const timeout = time.Second * 5
+		By("Setting up a new reconciler")
+		act := newTestActuator()
+		reconciler = newReconciler(mgr, act)
+
+		err = add(mgr, reconciler, "testing")
+		Expect(err).NotTo(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_test.go:62]: More concise?
```suggestion
		Expect(add(mgr, reconciler, "testing")).To(Succeed())
```

File: pkg/controller/machine/machine_controller_test.go [hunk 16]
```diff
@@ -17,87 +17,217 @@ limitations under the License.
...
+		})).Should(Succeed())
+
+		Eventually(func() (machinev1.MachineStatus, error) {
+			machine := &machinev1.Machine{}
+			err := k8sClient.Get(ctx, objectKey(instance), machine)
+			if err != nil {
+				return machinev1.MachineStatus{}, err
+			}
+			return machine.Status, nil
+		}, timeout).ShouldNot(Equal(machinev1.MachineStatus{}))
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_test.go:126]: I think you could just do
```suggestion
		Eventually(k.Get(instance), timeout).Should(HaveField("Status", Not(Equal(machinev1.MachineStatus{}))))
```

File: pkg/controller/machine/machine_controller_test.go [hunk 17]
```diff
@@ -17,87 +17,217 @@ limitations under the License.
...
+		// // Expect platform status is not empty. This (check) means we've called the
+		// // actuator
+		// Eventually(func() (runtime.RawExtension, error) {
+		// 	machine := &machinev1.Machine{}
+		// 	err := k8sClient.Get(ctx, objectKey(instance), machine)
+		// 	if err != nil {
+		// 		return runtime.RawExtension{}, err
+		// 	}
+		// 	return *machine.Status.ProviderStatus, nil
+		// }, timeout).ShouldNot(Equal(runtime.RawExtension{}))
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_test.go:137]: As above you could use the `k.Get` and `HaveField` to make this more succinct

File: pkg/controller/machine/machine_controller_test.go [hunk 18]
```diff
@@ -17,87 +17,217 @@ limitations under the License.
...
+			instance.Status.AuthoritativeAPI = machinev1.MachineAuthorityClusterAPI
+		})).Should(Succeed())
+
+		By("Verifying that the AuthoritativeAPI is set to Cluster API")
+		Eventually(func() (machinev1.MachineAuthority, error) {
+			if err := k8sClient.Get(ctx, objectKey(instance), instance); err != nil {
+				return "", err
+			}
+			return instance.Status.AuthoritativeAPI, nil
+		}, timeout).Should(Equal(machinev1.MachineAuthorityClusterAPI))
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_test.go:175]: Could use `k.Get` and `HaveField` here as well, you can do `HaveField("Status.MachineAuthority", ...)` and change the fields as well

File: pkg/controller/machine/machine_controller_test.go [hunk 19]
```diff
@@ -17,87 +17,217 @@ limitations under the License.
...
+
+		By("Verifying the paused condition is approproately set to true")
+		Eventually(func() ([]machinev1.Condition, error) {
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_test.go:178]: `k.Get`, `HaveField("Status.Conditions", ContainElement(...))`

File: pkg/controller/machine/machine_controller_test.go [hunk 20]
```diff
@@ -17,87 +17,217 @@ limitations under the License.
...
+}
+
+func cleanResources() error {
+	machines := &machinev1.MachineList{}
+	if err := k8sClient.List(ctx, machines); err != nil {
+		return err
+	}
+	for _, machine := range machines.Items {
+		m := machine
+		if err := k8sClient.Delete(ctx, &m); err != nil {
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_test.go:227]: I believe there's a DeleteAll call you can do

Or try the cleanup utils on the actuator package side

File: pkg/controller/machine/controller.go (hunk 12 above)
Comment (theobarberbany) [reply pkg/controller/machine/controller.go:79]: Hmm - I'm using the same reasons and condition across the machine and machineset controllers. I could always put them somewhere common, although I'm not sure where's best to do that :)

Comment (JoelSpeed) [reply pkg/controller/machine/controller.go:79]: That's ok, maybe just group them together somehow and add a comment explaining why they are public

File: pkg/controller/machine/machine_controller_suite_test.go (hunk 13 above)
Comment (theobarberbany) [reply pkg/controller/machine/machine_controller_suite_test.go:45]: This is copypasta from the machineset controller - but yes, that makes more sense!

File: pkg/controller/machine/controller.go (hunk 1 above)
Comment (theobarberbany) [reply pkg/controller/machine/controller.go:178]: Yep not done this yet - going to feature gate + this once tests are done :D

File: pkg/controller/machine/machine_controller_suite_test.go (hunk 13 above)
Comment (theobarberbany) [reply pkg/controller/machine/machine_controller_suite_test.go:45]: Ah - it's so the non ginkgo tests don't fail :/

Comment (JoelSpeed) [reply pkg/controller/machine/machine_controller_suite_test.go:45]: In that case, we can fix this up at some other point

Comment (theobarberbany) [reply pkg/controller/machine/machine_controller_suite_test.go]: Have added `TestMain` with error checking and not moved to `BeforeEach()`

File: pkg/controller/machine/controller.go (hunk 3 above)
Comment (theobarberbany) [reply pkg/controller/machine/controller.go:205]: check every path where we exit the reconcile loop (after where we call updatestatus) we dont need the additional one when setting false

File: pkg/controller/machine/controller.go [hunk 21]
```diff
@@ -89,9 +91,13 @@ func AddWithActuator(mgr manager.Manager, actuator Actuator) error {
 	return AddWithActuatorOpts(mgr, actuator, controller.Options{})
 }
 
-func AddWithActuatorOpts(mgr manager.Manager, actuator Actuator, opts controller.Options) error {
+func AddWithActuatorAndFeatureGates(mgr manager.Manager, actuator Actuator, gate featuregate.MutableFeatureGate) error {
+	return AddWithActuatorOpts(mgr, actuator, controller.Options{}, gate)
+}
+
+func AddWithActuatorOpts(mgr manager.Manager, actuator Actuator, opts controller.Options, gate ...featuregate.MutableFeatureGate) error {
```
Comment (theobarberbany) [inline pkg/controller/machine/controller.go:98]: I think this should probably just be removed. We don't want some instances where we can add an operand with feature gates, and then one without.

Comment (theobarberbany) [reply pkg/controller/machine/controller.go:98]: It also makes the contract less fragile (`gate[0]` below)

File: pkg/controller/controller.go [hunk 22]
```diff
@@ -13,3 +14,13 @@ func AddToManager(m manager.Manager, opts manager.Options, fnList ...func(manage
 	}
 	return nil
 }
+
+// AddToManagerWithFeatureGate adds all Controllers to the Manager along with a feature gate accessor
+func AddToManagerWithFeatureGate(m manager.Manager, opts manager.Options, featureGateAccessor featuregates.FeatureGateAccess, fnList ...func(manager.Manager, manager.Options, featuregates.FeatureGateAccess) error) error {
```
Comment (JoelSpeed) [inline pkg/controller/controller.go:19]: I think on this occasion I would probably just extend the existing AddToManager rather thann introducing a new one, this isn't shared outside this repo is it?

File: pkg/controller/machine/controller.go [hunk 23]
```diff
@@ -167,6 +185,40 @@ func (r *ReconcileMachine) Reconcile(ctx context.Context, request reconcile.Requ
...
+			conditions.Set(m, conditions.TrueConditionWithReason(
+				PausedCondition,
+				PausedConditionReason,
+				"The AuthoritativeAPI is set to %s", m.Status.AuthoritativeAPI,
+			))
+			if patchErr := r.updateStatus(ctx, m, ptr.Deref(m.Status.Phase, ""), nil, originalConditions); patchErr != nil {
+				klog.Errorf("%v: error patching status: %v", machineName, patchErr)
+			}
+
+			klog.Infof("%v: setting paused to true and returning early", machineName)
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller.go:205]: Might suggest rewording this since end users probably don't care about returning early
```suggestion
			klog.Infof("%v: machine is paused, taking no further action", machineName)
```

File: pkg/controller/machine/controller.go [hunk 24]
```diff
@@ -167,6 +185,40 @@ func (r *ReconcileMachine) Reconcile(ctx context.Context, request reconcile.Requ
...
+			NotPausedConditionReason,
+			machinev1.ConditionSeverityInfo,
+			"The AuthoritativeAPI is set to %s", m.Status.AuthoritativeAPI,
+		))
+		if patchErr := r.updateStatus(ctx, m, ptr.Deref(m.Status.Phase, ""), nil, originalConditions); patchErr != nil {
+			klog.Errorf("%v: error patching status: %v", machineName, patchErr)
+		}
+	}
+
+	klog.Infof("machine %v starting validateMachine", machineName)
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller.go:221]: Either make this more end user facing or drop, personally I think this may be noisy

File: pkg/controller/machine/machine_controller_suite_test.go (hunk 14 above)
Comment (JoelSpeed) [reply pkg/controller/machine/machine_controller_suite_test.go:99]: Bump, don't think this one has been looked at yet

File: pkg/controller/machine/machine_controller_test.go [hunk 25]
```diff
@@ -17,87 +17,243 @@ limitations under the License.
...
+		Eventually(k.UpdateStatus(instance, func() {
+			instance.Status.AuthoritativeAPI = machinev1.MachineAuthorityMachineAPI
+		})).Should(Succeed())
+
+		Eventually(k.Object(instance), timeout).Should(HaveField("Status", Not(Equal(machinev1.MachineStatus{}))))
+
+		// // Expect platform status is not empty. This (check) means we've called the
+		// // actuator. Given our test actuator doesn't set any status, we may need to expect to be called.
+
+		// Eventually(k.Object(instance), timeout).Should(HaveField("Status.ProviderStatus", Not(Equal(runtime.RawExtension{}))))
```
Comment (JoelSpeed) [inline pkg/controller/machine/machine_controller_test.go:129]: This is todo before merge?

File: pkg/controller/machine/machine_controller_test.go (hunk 20 above)
Comment (JoelSpeed) [reply pkg/controller/machine/machine_controller_test.go:235]: Bump, still to look at

File: pkg/controller/machineset/controller.go [hunk 26]
```diff
@@ -164,6 +180,49 @@ func (r *ReconcileMachineSet) Reconcile(ctx context.Context, request reconcile.R
...
+				PausedConditionReason,
+				"The AuthoritativeAPI is set to %s", machineSet.Status.AuthoritativeAPI,
+			))
+
+			_, err := updateMachineSetStatus(r.Client, machineSet, machineSetCopy.Status)
+			if err != nil {
+				klog.Errorf("%v: error updating status: %v", machineSet.Name, err)
+			}
+
+			klog.Infof("%v: setting paused to true and returning early", machineSet.Name)
```
Comment (JoelSpeed) [inline pkg/controller/machineset/controller.go:210]: Same feedback as the other one, make a little more user facing

File: pkg/operator/sync.go [hunk 27]
```diff
@@ -635,6 +637,16 @@ func getProxyArgs(config *OperatorConfig) []corev1.EnvVar {
...
 }
 
+// buildFeatureGatesString builds a string with the format: --feature-gates=<name>=<bool>,<name>=<bool>...
+func buildFeatureGatesString(featureGates map[string]bool) string {
+	var parts []string
+	for name, enabled := range featureGates {
+		part := fmt.Sprintf("%s=%t", name, enabled)
+		parts = append(parts, part)
+	}
+	return "--feature-gates=" + strings.Join(parts, ",")
```
Comment (JoelSpeed) [inline pkg/operator/sync.go:647]: When we merge this, will it then break all of the operands? If they are erroring on unknown flags we will have to add the flag to the operand first (doing nothing?) and then merge this

**Review by JoelSpeed — COMMENTED** [review]
Looks like this is nearly ready, some nits, a couple of outstanding bigger points

File: pkg/controller/machine/machine_controller_test.go (hunk 25 above)
Comment (theobarberbany) [reply pkg/controller/machine/machine_controller_test.go:129]: I think we should leave this for another time. The current test actuator only counts calls, it doesn't actually do anything like setting a status.

File: pkg/controller/machine/controller.go [hunk 28]
```diff
@@ -89,9 +91,13 @@ func AddWithActuator(mgr manager.Manager, actuator Actuator) error {
...
 
-func AddWithActuatorOpts(mgr manager.Manager, actuator Actuator, opts controller.Options) error {
+func AddWithActuatorAndFeatureGates(mgr manager.Manager, actuator Actuator, gate featuregate.MutableFeatureGate) error {
+	return AddWithActuatorOpts(mgr, actuator, controller.Options{}, gate)
+}
+
+func AddWithActuatorOpts(mgr manager.Manager, actuator Actuator, opts controller.Options, gate ...featuregate.MutableFeatureGate) error {
 	machineControllerOpts := opts
-	machineControllerOpts.Reconciler = newReconciler(mgr, actuator)
+	machineControllerOpts.Reconciler = newReconciler(mgr, actuator, gate[0])
```
[the remark this answers is not in the document]
Comment (theobarberbany) [reply pkg/controller/machine/controller.go:100]: I think what i was trying to do here doesn't make sense any way. As discussed in the other comment, will have only `AddWithActuator`

File: pkg/controller/controller.go (hunk 22 above)
Comment (theobarberbany) [reply pkg/controller/controller.go]: I've reverted this back to `AddToManagerWithFeatureGates` as we share it across controllers without the paused condition currently, eg the machine healthcheck & nodelink controllers.

File: pkg/controller/machine/controller.go (hunk 3 above)
Comment (theobarberbany) [reply pkg/controller/machine/controller.go]: I don't think this will work without some rejigging - there are a number of places where we exit without calling updateStatus. 

We could just return after the inital call to set paused=false, but that will mean a requeue.

File: pkg/operator/sync.go (hunk 27 above)
Comment (theobarberbany) [reply pkg/operator/sync.go:647]: I think this is the case (e.g looking at machine-api-provider-aws) will fail with an error like: `flag provided but not defined: -feature-gates`

File: pkg/controller/machine/controller.go [hunk 29]
```diff
@@ -264,6 +314,7 @@ func (r *ReconcileMachine) Reconcile(ctx context.Context, request reconcile.Requ
 		return reconcile.Result{}, nil
 	}
 
+	// Exits without calling updateStatus
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller.go:317]: Is this bad?

File: pkg/controller/machine/controller_test.go [hunk 30]
```diff
@@ -681,9 +715,16 @@ func TestUpdateStatus(t *testing.T) {
...
+	g := NewWithT(t)
+	_, testEnv, err := StartEnvTest()
+	g.Expect(err).ToNot(HaveOccurred())
+	defer func() {
+		g.Expect(testEnv.Stop()).To(Succeed())
+	}()
+
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			g := NewWithT(t)
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller_test.go:686]: Removing this means you now fail the entire suite instead of just one test case, notice the scoping of the `t` variable here vs moving this to line 719. You can just have it on both lines, or, avoid shadowing by calling the subtest `g` something else, `gs` is commonly used

File: pkg/controller/machine/machine_controller_test.go (hunk 25 above)
Comment (JoelSpeed) [reply pkg/controller/machine/machine_controller_test.go:130]: Probably want some explanation of that then I guess?

File: pkg/controller/machineset/controller.go [hunk 31]
```diff
@@ -164,6 +180,49 @@ func (r *ReconcileMachineSet) Reconcile(ctx context.Context, request reconcile.R
...
 
+	var currentFeatureGates featuregates.FeatureGate
+	currentFeatureGates, err := r.featureGateAccessor.CurrentFeatureGates()
+
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to retrieve current feature gates: %w", err)
+	}
+
+	if currentFeatureGates.Enabled(features.FeatureGateMachineAPIMigration) {
+		machineSetCopy := machineSet.DeepCopy()
```
Comment (JoelSpeed) [inline pkg/controller/machineset/controller.go:191]: Do we need a copy here?

File: pkg/controller/machineset/machineset_controller_test.go [hunk 32]
```diff
@@ -131,30 +149,133 @@ var _ = Describe("MachineSet Reconciler", func() {
...
-		m := machine
-		if err := k8sClient.Delete(ctx, &m); err != nil {
-			return err
-		}
-	}
 
 	return nil
 }
+
+func objectKey(object metav1.Object) client.ObjectKey {
```
Comment (JoelSpeed) [inline pkg/controller/machineset/machineset_controller_test.go:258]: I think the client package has a function that does this already that you could use

File: pkg/controller/machine/controller_test.go (hunk 30 above)
Comment (theobarberbany) [reply pkg/controller/machine/controller_test.go:686]: Ah ok - will go the `gs` route!

File: pkg/controller/machineset/controller.go (hunk 31 above)
Comment (theobarberbany) [reply pkg/controller/machineset/controller.go:191]: yep my bad - not sure what I was thinking here!

Comment (theobarberbany) [reply pkg/controller/machineset/controller.go:191]: Although, weirdly removing it is making the tests fail......

File: cmd/vsphere/main.go [hunk 33]
```diff
@@ -203,7 +206,25 @@ func main() {
...
+	// Sets up machine api migration feature gates
+	defaultMutableGate := feature.DefaultMutableFeatureGate
+	// TODO: Do we want selfmanaged, or all in here?
+	_, err = features.NewFeatureGateOptions(defaultMutableGate, apifeatures.SelfManaged, apifeatures.FeatureGateMachineAPIMigration)
+	if err != nil {
+		klog.Fatalf("Error setting up feature gates: %v", err)
+	}
+
+	featureGateMap := map[string]bool{
+		string(apifeatures.FeatureGateMachineAPIMigration): featureGates.Enabled(apifeatures.FeatureGateMachineAPIMigration),
```
[the remark this answers is not in the document]
Comment (theobarberbany) [reply cmd/vsphere/main.go:218]: Moving this to be in a different PR: https://github.com/openshift/machine-api-operator/pull/1269
